zmq_actor.hpp

```diff
@@ -0,0 +1,120 @@
+/*
+    Copyright (c) 2020 ZeroMQ community
+    Copyright (c) 2020 Brett Viren
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"), to
+    deal in the Software without restriction, including without limitation the
+    rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+    sell copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+*/
+
+#ifndef __ZMQ_ACTOR_HPP_INCLUDED__
+#define __ZMQ_ACTOR_HPP_INCLUDED__
+
+#include "zmq.hpp"
+
+#ifdef ZMQ_CPP11
+// Actor needs some threading.
+// Initial implementation uses std::thread so C++11 only.
+#include <thread>
+
+namespace zmq {
+
+    typedef std::pair<zmq::socket_t, zmq::socket_t> pipe;
```

Comment:
The nomenclature is a bit confusing, here pipe is a pair of two sockets, but in actor_t pipe() is a single socket.

Reply:
I agree, it is imprecise.  I've seen "pipe" refer to the pair and to either end.  More often it's the latter, but that's also imprecise as "pipe" doesn't communicate well which end is being referred to.  I am not thinking of any better terms so I keep consistency with CZMQ.  Maybe `pipe_ends`.  Suggestions?

Reply:
Pipes imply one directional flow of data (http://api.zeromq.org/2-1:zmq-socket#toc11), so I think that is not an optimal name (same as channel).

So maybe you could not provide this typedef at all, and instead call the creation function `create_socket_pair` or `create_paired_sockets`. I'm not sure about the `pipe()` function, maybe just `get_socket()`.

Reply:
I use "link" and "linked pairs" nomenclature.